Fix #5001: Set Up Firestore and Upload Free Form Responses

adhiamboperes opened the pull request:
## Explanation
Fixes #5001.
This is PR 5 of 6 Planned PRs.

We are adding a whole new infrastructure for uploading data to Firebase Firestore.

### Key Changes
- Introducing a `FirestoreDataController` that handles Anonymous Authentication and logging to Firestore, based on connectivity status.
- Adding the Firestore, Auth and AppCheck dependencies, and reppining these maven dependencies.
- New event logger for converting event logs to documents(hashmaps) and uploading them to firestore.
- Introduce AuthenticationController
This new controller provides functionality to retrieve the current signed in user and also sign in a user anonymously, and can be called from anywhere. It has a fake that is to be used in tests.
- Adding the newly created `utility` module files to `CODEOWNERS`
- Creating wrappers for FirebaseAuth, and Firestore in order to allow switching out with fakes in tests and development versions using dagger.

The implementation relies on an AuthenticationWrapper to switch out the real and fake authentication controllers for testing purposes. To do this, I created a dagger module to provide the correct listener where needed, hence the test file changes in the _app_ module tests.

### Technical Decisions
- I opted to create a new cache for events that will eventually be uploaded to Firestore. This means they will not be accidentally uploaded to Firebase.
- I opted to use EventLogs and convert them to Documents on the fly as opposed to creating an entirely new model and logging infrastruture that is Firestore-specific.
- I calculated the `FirestoreLogStorageCacheSize` based on the [average length of a Google Review ](https://www.meetsoci.com/resources/insights/brands-franchises/state-of-google-reviews/#:~:text=Average%20review%20length%20ranges%20from,per%20review%20for%20department%20stores.&text=In%20comparison%20to%20other%20local%20platforms%2C%20Google%20reviews%20skew%20positively.)and the [storage size computation specs](https://firebase.google.com/docs/firestore/storage-size) defined by Firebase. I computed it based on 10 profiles on a device over a 3 month period.
- I opted to use existing log upload infrastructure to upload to firebase, as opposed to creating a whole new upload workflow so I can reuse the existing test infrastructure as well.

### Dependency Updates
- I added dependencies for Firestore and Auth. The selected versions were both the earliest compatible with the rest of our existing dependencies and the latest that didn't cause proguarding issues. There were incidental updates to transitive dependencies, both major and minor.

### Proguard
Added new -dontwarns for the following:
`-dontwarn javax.naming.**`  [Reference](https://github.com/firebase/firebase-android-sdk/blob/00d4626/firebase-firestore/proguard.txt.)

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## Upload Screenshots
### Firestore Upload
<img width="1130" alt="Screenshot 2024-01-23 at 11 18 26" src="https://github.com/oppia/oppia-android/assets/59600948/c555c283-06e5-471d-9084-b9a3b30f8592">

## Querying Synced Firestore Data on BigQuery
<img width="1071" alt="Screenshot 2024-01-23 at 11 26 26" src="https://github.com/oppia/oppia-android/assets/59600948/3ba78102-5521-4cb7-bd69-6ede5d777e7d">

<img width="1071" alt="Screenshot 2024-01-23 at 11 26 44" src="https://github.com/oppia/oppia-android/assets/59600948/46579afe-3c7b-4d5d-8bcf-a4e133a5f609">

adhiamboperes committed:
Add show survey popup event log

adhiamboperes committed:
Add begin survey event log

adhiamboperes committed:
Log survey response events

adhiamboperes committed:
Refactor tests

Some of the production code was updated in the previous commit.

adhiamboperes committed:
Add tests to verify the abandon survey event log

adhiamboperes committed:
Add tests to verify the mandatory questions event logs

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into nps-survey-events

adhiamboperes committed:
Add test for begin survey event

adhiamboperes committed:
Add tests for SurveyEventsLogger

adhiamboperes committed:
Add tests for Survey Events in OppiaLogger

adhiamboperes committed:
Add bazel BUILD definition for SurveyEventsLogger

adhiamboperes committed:
Fix nits

adhiamboperes committed:
Remove large tests that can't run on either local or CI.

These scenarios are covered in other tests so I'm fine with not trying to make them work.

adhiamboperes committed:
Bazel BUILD dependency

adhiamboperes committed:
Refactor TODO and semantics

adhiamboperes committed:
Update boolean naming

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into nps-survey-events

adhiamboperes committed:
Log optional survey response

This commit sets up the logging infrastructure, including tests.

Since Firestore is introduced, some tests in the app module may be broken, and will be fixed in the next commit.

This commit soes not handle actual upload to firestore yet, because firebase auth is not yet properly configured.

adhiamboperes committed:
Add firebase Auth fore uploading to firestore

adhiamboperes committed:
Add BUILD definitions for firestore loggers

adhiamboperes committed:
Add logic for syncing to firestore

adhiamboperes committed:
update maven dependencies

adhiamboperes committed:
Fix import ordering

adhiamboperes committed:
pull develop and resolve conflicts

adhiamboperes committed:
Add new files to CODEOWNERS

adhiamboperes committed:
Fix prod bazel dependencies

adhiamboperes committed:
Fix npe in domain/src/test/java/org/oppia/android/domain/oppialogger/loguploader/LogUploadWorkerTest.kt

seanlip commented on app/src/main/java/org/oppia/android/app/application/AbstractOppiaApplication.kt:
```diff
@@ -47,6 +50,8 @@ abstract class AbstractOppiaApplication(
     // TODO(#4751): Re-enable WorkManager for S+.
     if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) {
       FirebaseApp.initializeApp(applicationContext)
+      Firebase.firestore
```
I don't know enough Kotlin, but I'm a bit surprised that these lines don't seem to be calling anything (shouldn't they have a "()" after them if they are functions)?

If they aren't functions then what do these lines do?

adhiamboperes replied:
In Kotlin getters use direct access syntax like this, so this returns the FirebaseFirestore instance of the default FirebaseApp.

adhiamboperes replied:
I suppose this means that I can call them directly at usage site instead of globally in this init block.

seanlip replied:
But where is the returned value being stored? I don't see any "val ... = ..." statement.

adhiamboperes replied:
I have moved these getters to their call site where they are properly assigned to variables.

seanlip commented on app/src/main/java/org/oppia/android/app/devoptions/vieweventlogs/ViewEventLogsViewModel.kt:
```diff
@@ -14,11 +15,14 @@ import javax.inject.Inject
 ) : ObservableViewModel() {
 
-  private val eventList = debugAnalyticsEventLogger.getEventList()
+  private val firebaseEvents = debugAnalyticsEventLogger.getEventList()
```
What's the difference between a "firebase event" and a "firestore event"?

Is this explained anywhere, in the code/wiki?

adhiamboperes replied:
Firestore events are eventlogs that are intended to be uploaded to Firestore, while Firebase events are event logs that are intended for upload to Firebase. 

> Is this explained anywhere, in the code/wiki?

More of implied via naming and KDocs? Since I have used this naming system in multiple files in this PR, I am not sure which would be a centralized place to add a general explanation.

seanlip replied:
But then should we rename "analytics event logger" here to "firebase event logger"?

adhiamboperes replied:
The prod implementation of the AnalyticsEventLogger interface is named FirebaseAnalyticsEventLogger already. The debug version that is used here does not make that distinction, because the debug implementation is generic.
I have renamed this "firebaseEvents" variable to analyticsEvents to reflect that, and kept the "firestoreEvents" variable name because it is descriptive of its use. I have also added comments.

seanlip reviewed (commented):
Hi @adhiamboperes, thanks for the PR. I don't really have much feedback! Just a couple of comments from looking at the files.

adhiamboperes committed:
Remove redundant firestore/auth initialization

adhiamboperes committed:
Introduce AuthenticationController

This new controller provides functionality to retrieve the current signed in user and also sign in user anonymously, and can be called from anywhere. This decouples authentication from a firestore-only use case.

The implementation relies on an AuthenticationListener to switch out the real and fake authentication controllers for testing purposes.

This commit has not included the Auth module in app module test files(Fragment/Acticity tests), hence these will fail to build with a dagger error. This is fixed in the subsequent commit.

adhiamboperes committed:
Fix missing KDOc

adhiamboperes committed:
Add new files to CODEOWNERS

adhiamboperes committed:
Fix BUILD file formatting

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into nps-optional-response-upload

adhiamboperes committed:
Fix failing log upload tests

adhiamboperes committed:
Add more tests for FirestoreDataController

Now that we can successfully mock firebase auth, we can test for upload, cache and sync scenarios.

adhiamboperes committed:
Add tests for FakeAuthenticationController.kt

adhiamboperes committed:
Add missing bazel dep for build flavors

adhiamboperes committed:
Fix static check failures

adhiamboperes committed:
Add TestAuthenticationModule to test files.

This is to allow the fake version of firebase

adhiamboperes committed:
Fix failing tests

adhiamboperes committed:
Rename variable in ViewEventLogsViewModel.kt

To be more descriptive of its purpose.

adhiamboperes committed:
Add missing dependency for e2e instrumentation module

adhiamboperes commented:
I have 2 sets of failing tests in this PR, which on investigation seem to be caused by dependency version problems with Firestore. I think the best way to do this is to generate a PR that is exclusively for adding the new dependencies so @kkmurerwa and I can debug the proguard failures. I suspect that we mostly just need to lower the version to an older one, but I have no way of determining which one will be compatible with our dependencies so it will be more of a try-build agains different versions.

seanlip commented on config/proguard/firebase-components-proguard-rules.pro:
```diff
@@ -1,6 +1,8 @@
 # Reference: https://github.com/firebase/firebase-android-sdk/blob/82b02af331/firebase-components/proguard.txt.
+# Reference: https://github.com/firebase/firebase-android-sdk/blob/master/firebase-firestore/proguard.txt.
```
Just a note -- what is on "master" could change. You might want to pin a specific commit as was done in the previous line.

adhiamboperes replied:
Thanks for catching this @seanlip, I have updated the url

seanlip reviewed (approved):
Looks good. Just one comment, but no blockers from me. Thanks @adhiamboperes!

adhiamboperes committed:
Fix proguarding

Lower the versions of Firestore and Auth.

The previously included versions were not compatible with older libraries currently in use within the project.

I also added a dontwarn to proguard for FirestoreChannel

adhiamboperes commented:
> I have 2 sets of failing tests in this PR, which on investigation seem to be caused by dependency version problems with Firestore. I think the best way to do this is to generate a PR that is exclusively for adding the new dependencies so @kkmurerwa and I can debug the proguard failures. I suspect that we mostly just need to lower the version to an older one, but I have no way of determining which one will be compatible with our dependencies so it will be more of a try-build agains different versions.

I was able to find and appropriately downgrade the Firebase Auth and Firestore dependencies. Proguard works fine now.

adhiamboperes committed:
Fix failing eventlog test

Switch out logger for controller which should swap debug vs real event loggers

adhiamboperes committed:
Fix failing tests

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into nps-optional-response-upload

adhiamboperes committed:
Exempt AuthenticationModule from tests

I have been unable to mock firebase for this test class since it is supposed to verify the real controller is returned, and not a mock.

adhiamboperes committed:
Ignore failing test

This test cannot pass due to the version of robolectric.

There is a conflict between Firestore's SQLite database connection and Robolectric's shadow implementation of SQLite. Firestore is trying to access an illegal connection pointer (pointer 52), which indicates a problem with the database connection.

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into nps-optional-response-upload

adhiamboperes committed:
Reformat @Ignore

adhiamboperes committed:
Refactor test

Allow real auth in tests for firestore

adhiamboperes committed:
Refactor event logs viewmodel to use a fake in tests

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into nps-optional-response-upload

adhiamboperes committed:
Add FakeAuthenticationControllerFactory

This is to swap out the real factory in tests

adhiamboperes committed:
Add test for AuthenticationModule

Ensure the production controller is always created in prod

adhiamboperes committed:
Fix static check test failures

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/options/OptionsActivityTest.kt:
```diff
@@ -86,7 +87,6 @@ import org.robolectric.annotation.Config
 import org.robolectric.annotation.LooperMode
 import javax.inject.Inject
 import javax.inject.Singleton
-
```
Please re-add missing newline used for vertical spacing (consistency).

adhiamboperes replied:
Done

BenHenning commented on domain/build.gradle:
```diff
@@ -93,9 +93,11 @@ dependencies {
       'com.google.protobuf:protobuf-javalite:3.17.3',
-      "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
+      "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version",
+      'org.mockito:mockito-core:2.19.0',
```
Production builds should never need to include Mockito (and in fact such cases often indicate a dependency that should go in testing/).

adhiamboperes replied:
I removed the dependency. It is already included in /testing

BenHenning replied:
It still seems present in the latest changes unless I'm missing something.

adhiamboperes replied:
Removed in latest changes

BenHenning commented on scripts/assets/maven_dependencies.textproto:
```diff
@@ -830,6 +928,16 @@ maven_dependency {
+  artifact_name: "org.codehaus.mojo:animal-sniffer-annotations:1.18"
+  artifact_version: "1.18"
+  license {
+    license_name: "The Apache Software License, Version 2.0"
```
Do you know why this is saying Apache? Per https://mvnrepository.com/artifact/org.codehaus.mojo/animal-sniffer-annotations/1.18 the dependency seems to be using MIT. I didn't check any of the other dependencies, but can you please make sure that all of their licenses are the correct ones per their GitHub repositories (where available) & Maven artifact listing?

I wonder if a fix is needed in the license extractor.

adhiamboperes replied:
It could have been an error on my part. I deleted it and re-ran the script, found that it required manual update. 
I checked the other licenses and they are all correct, so I don't have a reason to think that there is a bug in the extractor.

BenHenning commented on domain/src/test/java/org/oppia/android/domain/auth/BUILD.bazel:
```diff
@@ -0,0 +1,31 @@
+"""
+Library for providing authentication with Firebase functionality in the app.
```
```suggestion
Tests for libraries providing authentication with Firebase functionality.
```

adhiamboperes replied:
Updated

BenHenning commented on utility/src/main/java/org/oppia/android/util/logging/firebase/BUILD.bazel:
```diff
@@ -82,8 +83,49 @@ kt_android_library(
 )
 
+kt_android_library(
+    name = "firestore_logger_impl",
```
Does this actually need to be accessed outside of this package? Normally implementations aren't directly accessed (and instead are indirectly depended on via modules).

adhiamboperes replied:
It can be depended on via this package's `prod_module`, so I have updated the visibility.

BenHenning commented on app/src/main/java/org/oppia/android/app/survey/SurveyOutroDialogFragmentPresenter.kt:
```diff
@@ -60,7 +60,7 @@ class SurveyOutroDialogFragmentPresenter @Inject constructor(
           is AsyncResult.Pending -> oppiaLogger.d("SurveyActivity", "Stopping survey session")
           is AsyncResult.Failure -> {
-            oppiaLogger.d("SurveyActivity", "Failed to stop the survey session")
+            oppiaLogger.d("SurveyActivity", "Failed to stop the survey session, ${it.error}")
```
You could pass the throwable, instead. That will ensure the stack trace is printed, as well as any chained cause throwables (the stringified version of a throwable isn't always a reliable summary of the problem).

adhiamboperes replied:
I'm not sure I understand - the [error](https://github.com/oppia/oppia-android/blob/08981272a8136b8da0be7c4197480790c36bf100/utility/src/main/java/org/oppia/android/util/data/AsyncResult.kt#L195) in AsyncResult.Failure is a throwable.

BenHenning replied:
``oppiaLogger.d()`` has a parameter for passing the throwable so that its stack trace is logged, too.

adhiamboperes replied:
Got it, I have modified this to use the overloaded function with the throwable.

BenHenning commented on config/proguard/firebase-components-proguard-rules.pro:
```diff
@@ -1,6 +1,9 @@
 
 -dontwarn com.google.firebase.components.Component$Instantiation
 -dontwarn com.google.firebase.components.Component$ComponentType
+-dontwarn com.google.firebase.firestore.remote.FirestoreChannel
```
I don't see FirestoreChannel listed in the new proguard link added above. What's the actual error being run into? It may be the case of incompatible dependencies, not a warning that can safely be ignored.

adhiamboperes replied:
The warning is:
```
Warning: com.google.firebase.firestore.remote.FirestoreChannel$1: can't find enclosing method 'void lambda$runBidiStreamingRpc$0$FirestoreChannel(com.google.firebase.firestore.remote.FirestoreChannel,io.grpc.ClientCall[],com.google.firebase.firestore.remote.IncomingStreamObserver,com.google.android.gms.tasks.Task)' in program class com.google.firebase.firestore.remote.FirestoreChannel
Warning: com.google.firebase.firestore.remote.FirestoreChannel$3: can't find enclosing method 'void lambda$runStreamingResponseRpc$1$FirestoreChannel(com.google.firebase.firestore.remote.FirestoreChannel,com.google.android.gms.tasks.TaskCompletionSource,java.lang.Object,com.google.android.gms.tasks.Task)' in program class com.google.firebase.firestore.remote.FirestoreChannel
Warning: com.google.firebase.firestore.remote.FirestoreChannel$4: can't find enclosing method 'void lambda$runRpc$2$FirestoreChannel(com.google.firebase.firestore.remote.FirestoreChannel,com.google.android.gms.tasks.TaskCompletionSource,java.lang.Object,com.google.android.gms.tasks.Task)' in program class com.google.firebase.firestore.remote.FirestoreChannel
```
I tried some fixes to resolve this but none of them work so far:
- Keeping firestore classes in proguard
- Declaring the `com.google.android.gms.tasks.TaskCompletionSource` dependency
- I couldn't find something related to `java.lang.Object`,

adhiamboperes replied:
This is now fixed

BenHenning commented on scripts/assets/test_file_exemptions.textproto:
```diff
@@ -645,6 +645,9 @@ exempted_file_path: "data/src/main/java/org/oppia/android/data/backends/gae/mode
 exempted_file_path: "data/src/main/java/org/oppia/android/data/backends/gae/model/GaeVoiceover.kt"
 exempted_file_path: "data/src/main/java/org/oppia/android/data/backends/gae/model/GaeWrittenTranslation.kt"
 exempted_file_path: "data/src/main/java/org/oppia/android/data/backends/gae/model/GaeWrittenTranslations.kt"
+exempted_file_path: "domain/src/main/java/org/oppia/android/domain/auth/AuthenticationController.kt"
```
This class has logic in it--it seems useful to test to help catch issues earlier in the pipeline.

adhiamboperes replied:
Since `AuthenticationController` is just a thin wrapper, I would fail  to test it directly because I don't have any business logic to test here. I am just mapping from task to completable future. And so not testing this class is a valid conclusion. I mean in terms of code or behaviour coverage, testing this class impacts my confidence very lightly. What I lean towards, is testing it as part of an integration test, in FirestoreDataController, where there is usage of it.

BenHenning replied:
I think of it a bit differently. The parent class is a controller independent of Firebase that has its own behaviors that must be fulfilled. This implementation needs to fulfill that.

A wrapper is an implementation-specific thing that just forwards calls and can be swapped out for a test-only fake.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/auth/FakeAuthenticationControllerFactory.kt:
```diff
@@ -0,0 +1,13 @@
+import javax.inject.Inject
+
+/** A test specific fake for the AuthenticationController.Factory. */
+class FakeAuthenticationControllerFactory @Inject constructor() {
```
Should this be implementing AuthenticationController.Factory?

Also, why do we need this? If it's not implementing the ``Factory`` implementation, is the main benefit just binding the mock to be reusable? The class doesn't seem to be providing any specific fake functionality.

adhiamboperes replied:
I removed this class since I refactored AuthenticationController, now it does not use the factory pattern.

BenHenning commented on scripts/assets/test_file_exemptions.textproto:
```diff
@@ -751,6 +754,7 @@ exempted_file_path: "testing/src/main/java/org/oppia/android/testing/OppiaTestAn
 exempted_file_path: "testing/src/main/java/org/oppia/android/testing/OppiaTestRule.kt"
 exempted_file_path: "testing/src/main/java/org/oppia/android/testing/OppiaTestRunner.kt"
 exempted_file_path: "testing/src/main/java/org/oppia/android/testing/RichTextViewMatcher.kt"
+exempted_file_path: "testing/src/main/java/org/oppia/android/testing/TestAuthenticationModule.kt"
```
Modules are usually worth testing to ensure the bindings stay intentional as the modules change over time.

adhiamboperes replied:
Added a test file and removed this exemption

BenHenning commented on scripts/assets/test_file_exemptions.textproto:
```diff
@@ -842,11 +846,14 @@ exempted_file_path: "utility/src/main/java/org/oppia/android/util/logging/Loggin
 exempted_file_path: "utility/src/main/java/org/oppia/android/util/logging/SyncStatusManager.kt"
 exempted_file_path: "utility/src/main/java/org/oppia/android/util/logging/SyncStatusModule.kt"
 exempted_file_path: "utility/src/main/java/org/oppia/android/util/logging/firebase/DebugAnalyticsEventLogger.kt"
+exempted_file_path: "utility/src/main/java/org/oppia/android/util/logging/firebase/DebugFirestoreEventLogger.kt"
```
Test doubles are always worth testing as failures they introduce can make tests that depend on them more difficult to test/verify. Debug implementations that can also show up in developer builds should be assumed as 'production-ready', so they also ought to have tests.

adhiamboperes replied:
Since I created an interface DebugFirestoreEventLogger, I will keep this here. Its implementation, DebugFirestoreEventLoggerImpl, has tests added.

BenHenning commented on scripts/assets/test_file_exemptions.textproto:
```diff
@@ -842,11 +846,14 @@ exempted_file_path: "utility/src/main/java/org/oppia/android/util/logging/Loggin
 exempted_file_path: "utility/src/main/java/org/oppia/android/util/logging/firebase/FirebaseLogUploader.kt"
 exempted_file_path: "utility/src/main/java/org/oppia/android/util/logging/firebase/FirebaseLogUploaderModule.kt"
+exempted_file_path: "utility/src/main/java/org/oppia/android/util/logging/firebase/FirestoreEventLogger.kt"
+exempted_file_path: "utility/src/main/java/org/oppia/android/util/logging/firebase/FirestoreEventLoggerProdImpl.kt"
```
Can tests be added for this implementation, too?

adhiamboperes replied:
As per the discussion, we will leave in this eemption for now.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/auth/AuthenticationListener.kt:
```diff
@@ -0,0 +1,14 @@
+import org.oppia.android.util.data.AsyncResult
+
+/** Listener for getting the authentication state of the default FirebaseApp. */
+interface AuthenticationListener {
```
Is this actually a listener? It seems to be providing functionality that must be called by users of implementations of this listener rather than being a conduit for information from elsewhere in the app (e.g. onUserSignedIn(), onSignedInAnonymously(), etc.). The dataflow directionality here seems reversed from what I'd normally expect of a listener.

adhiamboperes replied:
It is more of a wrapper, and I have renamed it as such.

BenHenning commented on utility/src/main/java/org/oppia/android/util/logging/firebase/FirestoreEventLogger.kt:
```diff
@@ -0,0 +1,16 @@
+  fun uploadEvent(eventLog: EventLog)
+
+  /** Returns the list of all [EventLog]s logged since the app opened. */
+  fun getEventList(): List<EventLog>
```
This doesn't seem like something we'd want to expose generally. Could we maybe have a separate interface, e.g. DebugFirestoreEventLogger that exposes this? We can bind the same implementation instances to both listener types in Dagger (such that the debug-only module provides an implementation for DebugFirestoreEventLogger that can be used by the debug-only view models).

adhiamboperes replied:
Done. I created a `DebugFirestoreEventLogger` with these two functions, and changed `FirestoreEventLogger` to have only the `uploadEvent` function.

BenHenning commented on testing/src/main/java/org/oppia/android/testing/FakeAuthenticationController.kt:
```diff
@@ -0,0 +1,38 @@
+
+/** A test specific fake for the AuthenticationController. */
+@Singleton
+class FakeAuthenticationController @Inject constructor() : AuthenticationListener {
```
Is this a fake AuthenticationController or AuthenticationListener? Regardless, it's generally best practice to fake at the highest level below 1p code (e.g. the Firebase dependencies, in this case). It's also best practice to use existing fakes where available rather than reimplementing them. However, if there aren't fakes for an external library, we can wrap that entire dependency and provide fakes for the thin layer that sits on top of the dependency, keeping all downstream code unfaked.

adhiamboperes replied:
It is a fake AuthenticationController, since its the test fake implementation of AuthenticationListener/Wrapper. I removed the mock for FirebaseUser from this class. I hope that this accurately answers this comment.

BenHenning reviewed (changes requested):
Thanks @adhiamboperes! I took a high-level pass on architecture and boilerplate. I didn't dig into the details on implementation or tests yet.

I think overall the approach looks good, but I'm hoping for some clarification in comment, code, or both for a few different parts of the approach. PTAL.

adhiamboperes committed:
Add newline after imports in OptionsActivityTest.kt

adhiamboperes committed:
Update file documentation

adhiamboperes committed:
Update animal-sniffer-annotations license

adhiamboperes committed:
Remove mockito from production deps

adhiamboperes committed:
Refactor auth tests build file comments

adhiamboperes committed:
Change firestore_logger_impl visibility to private

adhiamboperes committed:
Add test for TestAuthenticationModule

And remove the test file exemption to ensure the bindings stay intentional as the modules change over time.

adhiamboperes committed:
Refactor AuthenticationController and its bindings.

Removed the factory pattern and converted the class to a singleton

adhiamboperes committed:
Renamed AuthenticationListener to AuthenticationWrapper.kt

I renamed it because the methods encapsulated here are required to be called by implementations of the interface.

adhiamboperes committed:
Add TestAuthenticationModuleTest to CODEOWNERS

adhiamboperes committed:
Create a debug interface for FirestoreEventLogger

We can bind the same implementation instances to both listener types in Dagger (such that the debug-only module provides an implementation for DebugFirestoreEventLogger that can be used by the debug-only view models).

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into nps-optional-response-upload

adhiamboperes committed:
Refactor DebugFirestoreEventLoggerImpl

Instead of Injecting the real logger's factory, injecting the interface allows for swapping out the implementation with a fake for testing purposes.

adhiamboperes committed:
Fix failing tests

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into nps-optional-response-upload

adhiamboperes committed:
Fix DebugFirestoreEventLoggerImplTest tests

adhiamboperes commented:
@BenHenning, could you PTAL at this? I have addressed the comments.

adhiamboperes committed:
Fix proguard failure.

Updated firestore dependency

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into nps-optional-response-upload

adhiamboperes committed:
Fix failing tests

BenHenning commented on utility/build.gradle:
```diff
@@ -89,6 +89,8 @@ dependencies {
       'com.google.firebase:firebase-analytics-ktx:17.5.0',
       'com.google.firebase:firebase-core:17.5.0',
       'com.google.firebase:firebase-crashlytics:17.0.0',
+      'com.google.firebase:firebase-firestore-ktx:21.5.0',
```
Is this intentionally a different version than the one used in Bazel? Prefer keeping all new dependencies aligned in versions to avoid any potential skew between builds (where possible).

adhiamboperes replied:
Updated the gradle versions to match the bazel ones.

BenHenning commented on third_party/maven_install.json:
```diff
@@ -1,16 +1,20 @@
             "androidx.core:core:1.0.1": "androidx.core:core:1.3.1",
             "androidx.recyclerview:recyclerview:1.0.0": "androidx.recyclerview:recyclerview:1.1.0",
             "androidx.test:core:1.0.0": "androidx.test:core:1.4.0",
+            "com.google.auto.value:auto-value-annotations:1.7.5": "com.google.auto.value:auto-value-annotations:1.8.1",
```
For this and the checker-qual dependency, since these are just being resolved to higher dependencies anyway, can we go ahead and depend on those exact versions in versions.bzl?

adhiamboperes replied:
Done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/auth/AuthenticationWrapper.kt:
```diff
@@ -0,0 +1,14 @@
+import org.oppia.android.util.data.AsyncResult
+
+/** Wrapper for providing authentication functionality. */
+interface AuthenticationWrapper {
```
Suggest this being ``AuthenticationController`` and calling the other one ``AuthenticationControllerImpl`` or ``FirebaseAuthenticationControllerImpl``.

adhiamboperes replied:
Not sure if this is still valid per latest changes?

adhiamboperes replied:
Not sure if this is still valid per latest chnages?

BenHenning commented on domain/src/main/java/org/oppia/android/domain/auth/AuthenticationWrapper.kt:
```diff
@@ -0,0 +1,14 @@
+/** Wrapper for providing authentication functionality. */
+interface AuthenticationWrapper {
+  /** Returns the current signed in user or null if there is no authenticated user. */
+  fun getCurrentSignedInUser(): FirebaseUser?
```
It would be ideal to abstract this so that we're not depending on Firebase types directly in the domain layer (in case we need to add onto or change the authentication system).

adhiamboperes replied:
Implemented a wrapper for FirebaseUser

BenHenning commented on domain/src/main/java/org/oppia/android/domain/auth/AuthenticationController.kt:
```diff
@@ -0,0 +1,33 @@
+class AuthenticationController @Inject constructor(
+  private val firebaseAuth: FirebaseAuth
+) : AuthenticationWrapper {
+  /** Returns the current signed in user or null if there is no authenticated user. */
```
Here & next method: no-need for KDocs on overrides.

adhiamboperes replied:
This is outdated due to implementation change.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/auth/AuthenticationController.kt:
```diff
@@ -0,0 +1,33 @@
+import javax.inject.Singleton
+
+/** Controller for signing in and retrieving a Firebase user. */
+@Singleton
```
This class doesn't have state, so it doesn't need to be scoped.

adhiamboperes replied:
Removed the Singleton annotation

BenHenning commented on domain/src/main/java/org/oppia/android/domain/auth/AuthenticationController.kt:
```diff
@@ -0,0 +1,33 @@
+
+  /** Returns the result of an authentication task. */
+  override fun signInAnonymously(): CompletableDeferred<AsyncResult<Any?>> {
+    val deferredResult = CompletableDeferred<AsyncResult<Any?>>()
```
Is there a timeout functionality possible here? Concerned about cases where something breaks add neither onSuccess/onFailure callbacks are called.

adhiamboperes replied:
One example of a scenario where on success/failure might not be called is a connection failure because firebase does not throw an exception in this case(it will keep retrying in the background). To mitigate against this, this method is never called unless a connection is established in `FirestoreDataController`

adhiamboperes replied:
This is now wrapped in a withTimeout of 3 seconds

BenHenning replied:
I suggest moving the timeout to a constant.

That being said, is 3 seconds enough time? Normally network timeouts are on the order of ~30 seconds, so just wanted to check.

adhiamboperes replied:
I updated the timeout to 30s and also moved it to a constant.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/auth/AuthenticationController.kt:
```diff
@@ -0,0 +1,33 @@
+  /** Returns the result of an authentication task. */
+  override fun signInAnonymously(): CompletableDeferred<AsyncResult<Any?>> {
+    val deferredResult = CompletableDeferred<AsyncResult<Any?>>()
+    firebaseAuth.signInAnonymously()
```
Should we have a check on whether the user is already signed in?

adhiamboperes replied:
I delegated this to the calling class, in this case the `FirestoreDataController`.

BenHenning commented on testing/src/main/java/org/oppia/android/testing/FakeAuthenticationController.kt:
```diff
@@ -0,0 +1,42 @@
+  /** Sets the current signed in user. */
+  fun setSignedInUser(firebaseUser: FirebaseUser) {
+    currentUser = firebaseUser
+  }
```
Could this maybe be consolidated to a sealed class state? Would ensure that the values always make sense together (e.g. "signInSuccessful = false && currentUser != null" probably doesn't make much sense).

adhiamboperes replied:
Done

BenHenning commented on testing/src/main/java/org/oppia/android/testing/FakeAuthenticationController.kt:
```diff
@@ -0,0 +1,42 @@
+
+/** A test specific fake for the AuthenticationController. */
+@Singleton
+class FakeAuthenticationController @Inject constructor() : AuthenticationWrapper {
```
I suspect per my other comments that we can use the real implementation and separately wrap exactly the Firebase calls.

If an interface isn't needed for the controller, we can have a single AuthenticationController class that uses a FirebaseAuthWrapper interface (with corresponding FirebaseAuthWrapperImpl and FakeFirebaseAuthWrapper classes).

adhiamboperes replied:
This is now implemented

BenHenning commented on domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/FirestoreDataController.kt:
```diff
@@ -0,0 +1,169 @@
+  }
+
+  /**
+   * Logs a high priority event defined by [eventContext] corresponding to time [timestamp].
```
'High priority' may not help here since this is the only choice of methods for logging an event.

adhiamboperes replied:
Modified to ` Logs an event defined by [eventContext] corresponding to time [timestamp].`

BenHenning commented on domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/FirestoreDataController.kt:
```diff
@@ -0,0 +1,169 @@
+  fun getEventLogStore(): DataProvider<OppiaEventLogs> = firestoreEventsStore
+
+  /** Removes the first log report that had been recorded for upload. */
+  fun removeFirstEventLogFromStore() {
```
Why is this public? It doesn't seem that this is something we'd want random app module classes to call (which is fair game since this is being established as part of the controller's public API).

adhiamboperes replied:
I originally had this as the function was called in loguploadworker to remove logs after they have been uploaded. I have made a change so that it is called after each log has been uploaded within this controller. WDYT about this?

BenHenning replied:
I think that's cleaner, but does that mean this can be private now?

adhiamboperes replied:
Yes, I changed access to private

BenHenning commented on domain/src/test/java/org/oppia/android/domain/oppialogger/analytics/FirestoreDataControllerTest.kt:
```diff
@@ -0,0 +1,465 @@
+
+    val eventLog = fakeFirestoreEventLogger.getMostRecentEvent()
+
+    EventLogSubject.assertThat(eventLog).hasTimestampThat().isEqualTo(TEST_TIMESTAMP)
```
Prefer directly importing assertThat.

adhiamboperes replied:
Updated.

adhiamboperes committed:
Merge branch 'develop' into nps-optional-response-upload

adhiamboperes committed:
Merge branch 'develop' into nps-optional-response-upload

BenHenning commented on app/src/main/java/org/oppia/android/app/devoptions/vieweventlogs/ViewEventLogsViewModel.kt:
```diff
@@ -14,11 +15,17 @@ import javax.inject.Inject
+  // Retrieves events from cache that are meant to be uploaded to Firebase Analytics.
+  private val analyticsEvents = debugAnalyticsEventLogger.getEventList()
+
+  private val eventList = analyticsEvents + firestoreEvents
```
Should these maybe be sorted by timestamp so that the two sets of events aren't concatenated?

adhiamboperes replied:
I concatenated these because they are 2 seperate lists from different loggers. But I have made a change such that I add the lists to `eventList` then sort by timestamp.

adhiamboperes committed:
Create firebase and firestore wrappers

BenHenning commented on domain/src/test/java/org/oppia/android/domain/oppialogger/loguploader/LogUploadWorkerTest.kt:
```diff
@@ -402,6 +413,95 @@ class LogUploadWorkerTest {
+    testCoroutineDispatchers.runCurrent()
+    val workInfo = workManager.getWorkInfoById(request.id)
+
+    // assertThat(workInfo.get().state).isEqualTo(WorkInfo.State.FAILED)
```
Please remove any commented out code included in the PR.

adhiamboperes replied:
Uncommented this assertion. I previously missed this.

adhiamboperes commented:
@BenHenning, PTAL.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/FirestoreDataController.kt:
```diff
@@ -0,0 +1,171 @@
+
+  /** Removes the first log report that had been recorded for upload. */
+  fun removeFirstEventLogFromStore() {
+    println("removing first event log from store")
```
Please make sure to remove debug lines the PR.

adhiamboperes replied:
Removed🤦‍♀️

BenHenning commented on domain/src/test/java/org/oppia/android/domain/auth/FirebaseAuthWrapperImplTest.kt:
```diff
@@ -0,0 +1,3 @@
+package org.oppia.android.domain.auth
+
+class FirebaseAuthWrapperImplTest
```
What's this file for?

adhiamboperes replied:
Tests for the prod impl of FirebaseAuthWrapper

BenHenning commented on utility/src/main/java/org/oppia/android/util/logging/firebase/DebugFirestoreInstanceWrapperImpl.kt:
```diff
@@ -0,0 +1,17 @@
+
+/**
+ * A test and debug fake for the [FirestoreInstanceWrapper]. This is also used in debug environments
+ * to make the debug implementation testable. [FirebaseFirestore] requires an instance of []FirebaseApp],
```
```suggestion
 * to make the debug implementation testable. [FirebaseFirestore] requires an instance of [FirebaseApp],
```

adhiamboperes replied:
Corrected

BenHenning commented on utility/src/main/java/org/oppia/android/util/logging/firebase/FirestoreEventLogger.kt:
```diff
@@ -0,0 +1,13 @@
+  /**
+   * Converts eventLogs to Firestore documents and uploads or save them on disk.
+   *
+   * @param eventLog which contains all the relevant data to be reported.
```
```suggestion
   * @param eventLog which contains all the relevant data to be reported
```

adhiamboperes replied:
Fixed.

BenHenning commented on utility/src/main/java/org/oppia/android/util/logging/firebase/FirestoreEventLoggerProdImpl.kt:
```diff
@@ -0,0 +1,32 @@
+      ?.collection("nps_survey_open_feedback")
+      ?.add(document)
+      ?.addOnSuccessListener {
+        println("upload successful")
```
println here to remove as well.

adhiamboperes replied:
Done

BenHenning commented on utility/src/test/java/org/oppia/android/util/logging/firebase/DebugFirestoreEventLoggerImplTest.kt:
```diff
@@ -0,0 +1,181 @@
+    debugFirestoreLoggerImpl.clearAllEvents()
+    val isListEmpty = debugFirestoreLoggerImpl.getEventList().isEmpty()
+
+    Truth.assertThat(isListEmpty).isTrue()
```
assertThat should be directly imported.

adhiamboperes replied:
Fixed this.

BenHenning reviewed (changes requested):
Thanks @adhiamboperes! I originally was going to LGTM this right away, but given that there's 24 comments I think it'd nice to take one more pass (please let me know if you have any concerns with this).

adhiamboperes committed:
Address review comments

The comment on timeout possibility and creating a sealed class for fakeauthenticationwrapperimp have not been addressed in this commit.

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into nps-optional-response-upload

adhiamboperes committed:
Update maven install

adhiamboperes commented:
@BenHenning, PTAL.

adhiamboperes committed:
Add tests for prod impl of firebase auth

adhiamboperes committed:
Remove singleton annotation from TestAuthenticationModule

adhiamboperes committed:
Add FakeFirebaseAuthInstanceWrapperImpl to CODEOWNERS

adhiamboperes committed:
Fix failing eventlogs tests

adhiamboperes committed:
Add test file exemption for FakeFirebaseAuthInstanceWrapperImpl

adhiamboperes committed:
Add withtimeout to authentication logic

adhiamboperes committed:
Resolve background dispatcher dependency

adhiamboperes committed:
Ensure debug firestore pushes events to the cloud

This is important so that we can verify uploads and also test appcheck

BenHenning commented on domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/FirestoreDataController.kt:
```diff
@@ -41,13 +41,19 @@ class FirestoreDataController @Inject constructor(
+      eventLogsToUpload.forEach { eventLog ->
+        authenticateAndUploadToFirestore(eventLog)
+      }.also {
+        removeFirstEventLogFromStore()
```
Won't this only remove one event after all have been iterated?

adhiamboperes replied:
Yes, I modified this logic to call `removeFirstEventLogFromStore()` on each iteration and added tests to verify it.

BenHenning commented on domain/src/test/java/org/oppia/android/domain/auth/FirebaseAuthWrapperImplTest.kt:
```diff
@@ -1,3 +1,119 @@
+    )
+    val user = firebaseAuthWrapperImpl.currentUser
+
+    Truth.assertThat(user).isNull()
```
assertThat should be directly imported.

adhiamboperes replied:
Done

BenHenning commented on domain/src/test/java/org/oppia/android/domain/auth/FirebaseAuthWrapperImplTest.kt:
```diff
@@ -1,3 +1,119 @@
+
+  @Test
+  fun testAuthWrapperImpl_getCurrentSignedInUser_userIsNotSignedIn_returnsNull() {
+    firebaseAuthWrapper.signInAnonymously(
+      onSuccess = {},
+      onFailure = {}
+    )
```
This is a bit confusing with respect to the test name--it seems like the test _is_ attempting to log in.

adhiamboperes replied:
Fixed this

BenHenning commented on testing/src/main/java/org/oppia/android/testing/FakeFirebaseAuthWrapperImpl.kt:
```diff
@@ -9,29 +9,41 @@ import javax.inject.Singleton
   }
+
+  /** Represents a faked authentication state. */
+  sealed class FakeAuthState {
```
Probably could just be an enum since the classes aren't providing unique logic.

adhiamboperes replied:
Update this to an enum

BenHenning commented on testing/src/test/java/org/oppia/android/testing/FakeFirebaseAuthWrapperImplTest.kt:
```diff
@@ -64,6 +56,21 @@ class FakeFirebaseAuthWrapperImplTest {
+  }
+
+  @Test
+  fun testFakeAuthWrapper_simulateSignInSuccess_returnsFakeAuthStateFailure() {
```
```suggestion
  fun testFakeAuthWrapper_simulateSignInFailure_returnsFakeAuthStateFailure() {
```

adhiamboperes replied:
Fixed, thanks!

BenHenning commented on app/src/main/java/org/oppia/android/app/devoptions/vieweventlogs/ViewEventLogsViewModel.kt:
```diff
@@ -29,8 +37,14 @@ class ViewEventLogsViewModel @Inject constructor(
-    }.reversed()
+    return eventList
+      .apply {
+        addAll(analyticsEvents)
+        addAll(firestoreEvents)
```
This is going to change the eventList.

If we're not going to change eventList, you probably don't even need it anymore. Instead, you could just do something like:

```kotlin
return (analyticsEvents + firestoreEvents).map { ... }.sortedByDescending { ... }
```

adhiamboperes replied:
I have made the change per your suggestion

BenHenning commented on domain/src/main/java/org/oppia/android/domain/auth/AuthenticationController.kt:
```diff
@@ -0,0 +1,47 @@
+      } catch (e: TimeoutCancellationException) {
+        // Handle timeout exception
+        deferredResult.complete(
+          AsyncResult.Failure(TimeoutException("Authentication attempt timed out"))
```
Suggest wrapping the original exception in something new, like an ``IllegalStateException``, to preserve the stack trace history.

adhiamboperes replied:
Done

BenHenning reviewed (approved):
Thanks @adhiamboperes! Just had a few follow-up comments. Overall the PR LGTM (so I've approved it as such), but feel free to assign me to take another pass on the final deltas if you'd like (I'm fine with you resolving my follow-up comments, as well).

adhiamboperes committed:
Address PR feedback

- refactor FakeAuthState from sealed to enum class
- refactor auth timeout to constant
- fix remove uploaded logs logic
- fix firebase authwrapper test set up and naming
- refactor event log viewmodel list mapping

adhiamboperes committed:
Merge branch 'develop' into nps-optional-response-upload

adhiamboperes commented:
@BenHenning, I have resolved all the conversations as I was able to verify the changes myself. This PR is ready to be merged, thanks!

adhiamboperes committed:
Merge branch 'develop' into nps-optional-response-upload
